# __Introduction to BERT and Transformers Library__
- BERT stands for Bidirectional Encoder Representations from Transformers.
- BERT is pre-trained on a large corpus of unlabeled text, including the entire Wikipedia (that's 2,500 million words!) and the Book Corpus (800 million words).
- BERT is based on the Transformer architecture.

## Steps to be followed:
1. Import the required libraries
2. Analyze the sentiment using the transformer pipeline
3. Create text generation
4. Create named entity recognition (NER)
5. Generate a masked language model using a model and a tokenizer

### Step 1: Import the required libraries
- The code from the transformers import pipeline allows for easy access to pre-trained models and simplified execution of NLP tasks using the transformers library.



In [ ]:
from transformers import pipeline

###Step 2: Analyze the sentiment using  transformer pipeline

- Import the pipeline function from the Transformers library, which enables easy access to pre-trained NLP models.
- The snippet creates a sentiment analysis pipeline using the pre-trained model and uses it to classify the sentiment of the input text, **I hate you.**
- The result, including the sentiment label and score, is then printed.

In [ ]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")

result = classifier("I hate you")[0]
print(f"label: {result['label']}, with score: {round(result['score'], 4)}")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

label: NEGATIVE, with score: 0.9991


- Perform sentiment analysis on the text **I love you.**
- Print the sentiment analysis result.

In [ ]:
result = classifier("I love you")[0]
print(f"label: {result['label']}, with score: {round(result['score'], 4)}")

label: POSITIVE, with score: 0.9999


**Observation**
- The sentiment analysis model is highly confident that the sentiment of the text **I love you** is positive, with a score of 0.9999.

### Step 3: Create text generation
- It creates a text-generation pipeline using the pipeline function from the Transformers library.
- It generates text starting with the provided prompt **As far as I am concerned, I will** using the text generation pipeline with a maximum length of 50 tokens and without sampling, which is deterministic output.
- The generated text is then printed.

In [ ]:
text_generator = pipeline("text-generation")
print(text_generator("As far as I am concerned, I will", max_length=50, do_sample=False))


No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d.
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/148 [00:00<?, ?it/s]

GPT2LMHeadModel LOAD REPORT from: openai-community/gpt2
Key                  | Status     |  | 
---------------------+------------+--+-
h.{0...11}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Passing `generation_config` together with generation-related arguments=({'max_length', 'do_sample'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[{'generated_text': 'As far as I am concerned, I will be the first to admit that I am not a fan of the idea of a "free market." I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I think that the idea of a free market is a bit of a stretch. I th

### Step 4: Create named entity recognition (NER)
- It creates a NER pipeline using the pipeline function from the Transformers library.

- It applies the NER pipeline to the provided sequence, which is a text containing named entities. The pipeline identifies and extracts named entities such as organization names **Hugging Face Inc.**, locations **New York City**, and others. The extracted entities are then printed.

In [ ]:
ner_pipe = pipeline("ner")
sequence = """Hugging Face Inc. is a company based in New york city. Manhattan bridge is visible from the window."""

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision 4c53496.
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/391 [00:00<?, ?it/s]

BertForTokenClassification LOAD REPORT from: dbmdz/bert-large-cased-finetuned-conll03-english
Key                      | Status     |  | 
-------------------------+------------+--+-
bert.pooler.dense.bias   | UNEXPECTED |  | 
bert.pooler.dense.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

- Print the entities after performing named entity recognition on the sequence.

In [ ]:
for entity in ner_pipe(sequence):
    print(entity)

{'entity': 'I-ORG', 'score': np.float32(0.99954754), 'index': 1, 'word': 'Hu', 'start': 0, 'end': 2}
{'entity': 'I-ORG', 'score': np.float32(0.99008596), 'index': 2, 'word': '##gging', 'start': 2, 'end': 7}
{'entity': 'I-ORG', 'score': np.float32(0.9977804), 'index': 3, 'word': 'Face', 'start': 8, 'end': 12}
{'entity': 'I-ORG', 'score': np.float32(0.99955183), 'index': 4, 'word': 'Inc', 'start': 13, 'end': 16}
{'entity': 'I-LOC', 'score': np.float32(0.99441695), 'index': 11, 'word': 'New', 'start': 40, 'end': 43}
{'entity': 'I-LOC', 'score': np.float32(0.86684954), 'index': 12, 'word': 'yo', 'start': 44, 'end': 46}
{'entity': 'I-LOC', 'score': np.float32(0.9872277), 'index': 13, 'word': '##rk', 'start': 46, 'end': 48}
{'entity': 'I-LOC', 'score': np.float32(0.9989699), 'index': 16, 'word': 'Manhattan', 'start': 55, 'end': 64}



### Step 5: Generate a masked language model using a model and a tokenizer

- Masked language modeling using a model and a tokenizer
  - Masked language modeling is a task where a model fills in masked tokens in a sequence, improving its understanding of language. It involves predicting missing tokens by considering the context of surrounding words.

- The process includes the following steps:
  - Instantiate a tokenizer and a model from the checkpoint name.
  - Define a sequence with a masked token, placing the tokenizer.mask_token instead of a word.
  - Encode that sequence into a list of IDs and find the position of the masked token in that list.
  - Retrieve the predictions at the index of the masked token.
  - Retrieve the top 5 tokens using the PyTorch topk or TensorFlow top_k methods.
  - Replace the masked token with the tokens and print the results.

 ### Masked language modeling
- Import the necessary modules from the transformer library and torch.
- Load the pre-trained tokenizer and model.
- Define the input sequence with a masked token.
- Tokenize the input sequence and convert it to tensors.
- Find the index of the masked token and generate token predictions using the model.
- Get the indices of the top 5 predicted tokens and print them in the sequence.
- Load the pre-trained tokenizer and model.
- Define the input sequence with a masked token.
- Tokenize the input sequence and convert it to tensors.
- Find the index of the masked token and generate token predictions using the model.
- Get the indices of the top 5 predicted tokens and print them in the sequence.
- Print the top 5 predicted tokens in the masked position.

In [ ]:
from transformers import AutoModelForMaskedLM, AutoTokenizer
import torch

tokenizer = AutoTokenizer.from_pretrained("distilbert-base-cased")
model = AutoModelForMaskedLM.from_pretrained("distilbert-base-cased")

sequence = (
    "Distilled models are smaller than the models they mimic. Using them instead of the large "
    f"versions would help {tokenizer.mask_token} our carbon footprint."
)

inputs = tokenizer(sequence, return_tensors="pt")
mask_token_index = torch.where(inputs["input_ids"] == tokenizer.mask_token_id)[1]

token_logits = model(**inputs).logits
mask_token_logits = token_logits[0, mask_token_index, :]

top_5_tokens = torch.topk(mask_token_logits, 5, dim=1).indices[0].tolist()

for token in top_5_tokens:
    print(sequence.replace(tokenizer.mask_token, tokenizer.decode([token])))

config.json:   0%|          | 0.00/465 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/263M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/105 [00:00<?, ?it/s]

Distilled models are smaller than the models they mimic. Using them instead of the large versions would help reduce our carbon footprint.
Distilled models are smaller than the models they mimic. Using them instead of the large versions would help increase our carbon footprint.
Distilled models are smaller than the models they mimic. Using them instead of the large versions would help decrease our carbon footprint.
Distilled models are smaller than the models they mimic. Using them instead of the large versions would help offset our carbon footprint.
Distilled models are smaller than the models they mimic. Using them instead of the large versions would help improve our carbon footprint.


**Observation**
- The output provides alternative sentence suggestions by replacing the masked token with different predicted tokens, demonstrating how using distilled models instead of larger ones can impact the carbon footprint.

In [ ]:
# =========================
# INSTALL DEPENDENCIES
# =========================
!pip install -q transformers torch

# =========================
# IMPORTS
# =========================
from transformers import pipeline
from typing import List, Dict

# =========================
# MODEL LOADER
# =========================
def load_bert_sentiment_model():
    """
    Load a pretrained BERT sentiment analysis pipeline.
    """
    classifier = pipeline(
        task="sentiment-analysis",
        model="nlptown/bert-base-multilingual-uncased-sentiment"
    )
    return classifier

# =========================
# SINGLE TEXT PREDICTION
# =========================
def predict_sentiment(model, text: str) -> Dict:
    """
    Predict sentiment for a single sentence.
    """
    result = model(text)[0]
    return {
        "text": text,
        "sentiment": result["label"],
        "confidence": round(result["score"], 4)
    }

# =========================
# BATCH PREDICTION
# =========================
def batch_predict(model, texts: List[str]) -> List[Dict]:
    """
    Predict sentiment for multiple texts.
    """
    results = model(texts)

    formatted_results = []
    for text, result in zip(texts, results):
        formatted_results.append({
            "text": text,
            "sentiment": result["label"],
            "confidence": round(result["score"], 4)
        })

    return formatted_results

# =========================
# MAIN EXECUTION
# =========================
def main():
    model = load_bert_sentiment_model()

    sample_texts = [
        "This course is absolutely amazing!",
        "The product quality is terrible.",
        "The meeting was okay, nothing special."
    ]

    print("=== SINGLE PREDICTION ===")
    single_result = predict_sentiment(model, sample_texts[0])
    print(single_result)

    print("\n=== BATCH PREDICTION ===")
    batch_results = batch_predict(model, sample_texts)

    for result in batch_results:
        print(result)

# Run
main()

config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

=== SINGLE PREDICTION ===
{'text': 'This course is absolutely amazing!', 'sentiment': '5 stars', 'confidence': 0.9522}

=== BATCH PREDICTION ===
{'text': 'This course is absolutely amazing!', 'sentiment': '5 stars', 'confidence': 0.9522}
{'text': 'The product quality is terrible.', 'sentiment': '1 star', 'confidence': 0.6487}
{'text': 'The meeting was okay, nothing special.', 'sentiment': '3 stars', 'confidence': 0.788}
